OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -0,0 +1,105 @@
+#region Copyright & License Information
+/*
+* Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+* This file is part of OpenRA, which is free software. It is made
+* available to you under the terms of the GNU General Public License
+* as published by the Free Software Foundation. For more information,
+* see COPYING.
+*/
+#endregion
+
+using System;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	class SandwormInfo : WandersInfo
+	{
+		public const string IdleSequenceName = "idle";
+		public const int TargetRescanIntervalInTicks = 5;
+
+		public override object Create(ActorInitializer init) { return new Sandworm(init.Self, this); }
+	}
+
+	class Sandworm : Wanders, ITick, INotifyBecomingIdle
+	{
+		readonly Lazy<IMove> move;
+		readonly Lazy<RenderUnit> renderUnit;
+		readonly Lazy<AttackBase> attackTrait;
+
+		int targetCountdown;
+		uint currentTargetActorId;
+
+		public Sandworm(Actor self, SandwormInfo info)
+			: base(self, info)
+		{
+			move = Exts.Lazy(self.Trait<IMove>);
+			renderUnit = Exts.Lazy(self.Trait<RenderUnit>);
+			attackTrait = Exts.Lazy(self.Trait<AttackBase>);
+		}
+
+		public new void OnBecomingIdle(Actor self)
+		{
+			if (renderUnit.Value.DefaultAnimation.CurrentSequence.Name != SandwormInfo.IdleSequenceName)
+				renderUnit.Value.DefaultAnimation.ReplaceAnim("idle");
+
+			base.OnBecomingIdle(self);
+		}
+
+		public override void DoAction(Actor self, CPos targetCell)
+		{
+			self.QueueActivity(move.Value.MoveWithinRange(Target.FromCell(self.World, targetCell, SubCell.Any), WRange.FromCells(1)));
+		}
+
+		public void Tick(Actor self)
+		{
+			if (--targetCountdown > 0)
+				return;
+
+			targetCountdown = SandwormInfo.TargetRescanIntervalInTicks;
+
+			var food = self.World.ActorsWithTrait<WormFood>();
+			var desiredPair = new TraitPair<WormFood>();
+			var maxStrength = 0.0;
+
+			foreach (var traitPair in food)
+			{
+				if (!traitPair.Actor.IsInWorld)
+					continue;
+
+				if (traitPair.Trait.CurrentNoise <= 0)
+					continue;
+
+				var distance = (self.Location - traitPair.Actor.Location).Length;
+				var strength = traitPair.Trait.CurrentNoise - traitPair.Trait.NoiseAttenuation * distance;
+
+				if (maxStrength >= strength)
+					continue;
+
+				maxStrength = strength;
+				desiredPair = traitPair;
+			}
+
+			// No target was found
+			if (maxStrength == 0.0)
+			{
+				currentTargetActorId = 0;
+				return;
+			}
+
+			Game.Debug("(Sandworm {0}) is attacking actor ({1} {2}).", self.ActorID, desiredPair.Actor.Info.Name, desiredPair.Actor.ActorID);
+			var dist = (self.Location - desiredPair.Actor.Location).Length;
+			var str = desiredPair.Trait.CurrentNoise - desiredPair.Trait.NoiseAttenuation * dist;
+			Game.Debug("Target Noise level is {0}. Distance is {1}, noise strength is {2}.", desiredPair.Trait.CurrentNoise, dist, str);
```

Comment:
Definitely. They are there because someone (I think it was @pchote) told me to drop some debug messages so the reviewers can see what's going on.